### ISSUE TITLE
Visualize sentence embedding (USE)

### ISSUE THREAD

**Issue by Sathibhu** [tensorflow/hub#227]
Hi,

I want to visualize the Universal Sentence Encoder's (USE) sentence embedding. 

1. What are the salient words in the encoded sentence? [InferSent from Facebook has a visualize method to do that].

2. Given two sentences and their corresponding vector embeddings from USE, how to visualize how each word in one sentence has an impact on the other sentence when a similarity between two sentences are calculated by dot product.

Thanks,
Sathish

**Comment by nataliedurgin** [comment]
I had the same desire, which prompted me to look at the InferSent code visualization method. 

[InferSent/models.py at master · facebookresearch/InferSent · GitHub](https://github.com/facebookresearch/InferSent/blob/master/models.py#L239)

Here are some of my thoughts, just in case they might be helpful. Disclaimer: All values used in the toy example below are complete nonsense.  Focus should be placed on the relative dimensions of the objects. 

Toy example of InferSent visualization:

Sentence
`The cat sat.` 

Vocabulary 
`[cat, sat]`

Suppose the “sentence embedding” space is 3-dimensional.  If I understand the InferSent model correctly (which I may not), it uses an LSTM, which is a RNN that at some point operates on a sentence, a sequence of words, with word embedding vectors as features of that input sequence. Suppose those word embedding vectors are, 

Word embedding vectors (say, 2-d):
- `cat: [1.0, 0.0]`
- `sat: [0.0, 1.0]`

Now, the model continues to build the (3-d) sentence embedding, but after all is said and done, to build the word importance, we go back to those word embedding vectors. InferSent has a transformation for those word vectors from their word embedding space into the sentence embedding space (the torch.nn.LSTM object) in a way that respects and leverages the word order and context in the sentence.  So we wind up with,

Word representations in the sentence embedding space: 
- `cat: [0.5, 0.4, 0.1]`
- `sat: [0.1, 0.3, 0.6]`

Then we simply tally the number of times each word has the  max value in each column.  “cat” has the max value in columns 1 and 2, and “sat” has the max value in column 3.  Then we just make a matplotlib bar chart that tells us that “cat” wins 2/3 of the time and “sat” wins 1/3 of the time. This gives us the relative importance. Simple, but effective for this architecture. 

The effectiveness of this method for computing word importance might carry to the USE architectures. In the USE paper, https://arxiv.org/pdf/1803.11175.pdf, we read,

“The context aware word representations are converted to a fixed length sentence encoding vector by computing the element-wise sum of the representations at each word position. The encoder takes as input a lower cased PTB tokenized string and outputs a 512 dimensional vector as the sentence embedding. (We then divide by the square root of the length of the sentence so that the differences between short sentences are not dominated by sentence length effects.)”

Sounds like in this case, the sentence encoding is built very directly from the word-embedding vectors, so as long as we can access those, we have the transformation and we would be good to attempt the InferSent visualization method.  For the other USE architecture we read ,

“…input embeddings for words and bi-grams are first averaged together and then passed through a feedforward deep neural network (DNN) to produce sentence embeddings” 

Again as long as we can access those input embeddings for words and bigrams, and as long as we can access the transformation, we would be go to try out the InferSent word importance method.  We would just need to figure out  how to access the correct analogue of the word embedding vectors for these USE model objects. I’m sure they are there, I just don’t know the syntax off-hand. You might beat me to it if you know Tensorflow well. 

I don’t know that this should be a feature like it is in InferSent, but it would be awesome to have an example for generating the necessary components of a Tensorboard visualization of the sentences  as points in a projection of the embedding space, with labels that are word importances computed in a way that is meaningful for the USE architectures.

**Comment by rmothukuru** [comment]
@Sathibhu ,
Can you please confirm if the response from @nataliedurgin has answered your question.

**Comment by rmothukuru** [comment]
Automatically closing due to lack of recent activity. Please update the issue when new information becomes available, and we will reopen the issue. Thanks!
